dafny/lcf/datalog/AstBuilder.cs

```diff
@@ -70,16 +70,88 @@ internal class AstBuilder : datalogBaseVisitor<object>
       return context.Start.Line;
     }
 
-    public override object VisitClause(datalogParser.ClauseContext context) {
-      var name = Sequence<char>.FromString(context.name.Text);
+    public override object VisitBuiltin(datalogParser.BuiltinContext context) {
       var terms = (Dafny.ISequence<_ITerm>) VisitTerm_list(context.term_list());
-      return new _module.Prop_App(name, terms);
+      switch (context.name.Text) {
+        case "split_string":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_SplitString(), terms);
+        case "sub_string":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_SubString(), terms);
+        case "string_lower":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_StringLower(), terms);
+        case "string_chars":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_StringChars(), terms);
+        case "length":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_Length(), terms);
+        case "lists:member":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Member(), terms);
+        case "member":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Member(), terms);
+        case "lists:reverse":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Reverse(), terms);
+        case "reverse":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Reverse(), terms);
+        case "lists:nth1":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Nth1(), terms);
+        case "nth1":
+           return new _module.Prop_BuiltinOp(new _module.Builtin_Nth1(), terms);
+        default:
+          return null;
+      }
+    }
+
+    public override object VisitExpression(datalogParser.ExpressionContext context) {
+      var left = (_module.Term)VisitTerm(context.left);
+      var right = (_module.Term)VisitTerm(context.right);
+      var terms = new List<_module.Term> { left, right };
+      var terms2 = (Dafny.ISequence<_ITerm>) Sequence<_module.Term>.Create(terms.Count, i => terms[(int) i]);
+      switch(context.name.Text) {
+        case "=<":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatLeq(), terms2);
+        case ">=":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatGeq(), terms2);
+        case "\\=":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatNeq(), terms2);
+        case "=\\=":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatNeq(), terms2);
+        case "<":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatLt(), terms2);
+        case ">":
+          return new _module.Prop_BuiltinOp(new _module.Builtin_NatGt(), terms2);
+        default:
+          return new _module.Prop_Eq(left, right);
```

Comment:
I think it would be better to have an explicit case for equals.

The default case should error out.

Reply:
Fixed